bitcoin/src/blockdata/transaction.rs

```diff
@@ -773,14 +774,29 @@ impl Decodable for Transaction {
 ///
 /// * `fee_rate` - the fee rate of the transaction being created.
 /// * `input_weight_prediction` - the predicted input weight.
+///
+/// # Returns
+///
+/// This will return [`NumOpResult::Error`] if the fee calculation (fee_rate * weight) overflows.
+/// Otherwise, [`NumOpResult::Valid`] will wrap the successful calculation.
 pub fn effective_value(
     fee_rate: FeeRate,
     input_weight_prediction: InputWeightPrediction,
     value: Amount,
-) -> Option<SignedAmount> {
+) -> NumOpResult<SignedAmount> {
     let weight = input_weight_prediction.total_weight();
-    let signed_input_fee = fee_rate.to_fee(weight).ok()?.to_signed();
-    value.to_signed().checked_sub(signed_input_fee)
+
+    let signed_input_fee = match fee_rate.to_fee(weight) {
+        NumOpResult::Valid(x) => x.to_signed(),
+        NumOpResult::Error(e) => return NumOpResult::Error(e)
+    };
+
+    // it's not possible to overflow when subtracted by the smallest given:
+    // Amount::MIN - SignedAmount::MAX = SignedAmount::MIN
+    let eff_val = value.to_signed().to_sat() - signed_input_fee.to_sat();
+    let signed_amount = SignedAmount::from_sat_i32(eff_val as i32);
+
+    NumOpResult::Valid(signed_amount)
 }
```

Comment:
This whole function can be written as:

```rust
    let weight = input_weight_prediction.total_weight();

    let fee = match fee_rate.to_fee(weight) {
        NumOpResult::Valid(x) => x.to_signed(),
        NumOpResult::Error(e) => return NumOpResult::Error(e)
    };
    value.to_signed() - fee
```

Reply:
> value.to_signed() - fee

This performs an unneeded `checked_sub()` though.

That's why I wrote the comment:

`
// it's not possible to overflow when subtracted by the smallest given:
// Amount::MIN - SignedAmount::MAX = SignedAmount::MIN
`

The reason why I'm avoiding an unneeded check is because checks are slow.

Reply:
Can you at least add a comment explaining why the cast is ok?

Reply:
done

Reply:
sorry, one need to edit one more time to correct an error

Reply:
done

Reply:
Thanks for making your reasoning explicit. Unfortunately it is not true. SignedAmount::MAX is about 2^51 which does not fit into an i32.

Reply:
> Thanks for making your reasoning explicit. Unfortunately it is not true. SignedAmount::MAX is about 2^51 which does not fit into an i32.

drat, I see that now.  And to use Amount::from_sat().unwrap() invokes a check, so maybe it's best to just do `value.to_signed() - fee` as Tobin suggested.

Reply:
Huh? What does this have to do with `i32`? I don't see it anywhere.

Reply:
> Huh? What does this have to do with i32

I was trying to use the constructor `pub const fn from_sat_i32(satoshi: i32)` as I recall.

The issue here is that the monadic types will do a check here even though we know it can't panic.  So in the commit history that's been erased, I was going from bitcoin types to native types (doing subtraction) and then back to bitcoin types to avoid the subtraction check.  The only way it's efficient to do so is to avoid the constructor checks, thus I was using `from_sat_i32`.

Reply:
I think I mentioned somewhere that we should have an infallible method for subtracting `Amount` from `Amount` and returning `Signed` amount.